R/grade_this.R

```diff
@@ -88,3 +88,151 @@ grade_this <- function(
     )
   }
 }
+
+#' Debug an exercise submission
+#' 
+#' When used in a `*-check` chunk or inside [grade_this()], `debug_this()`
+#' displays in the \pkg{learnr} tutorial a complete listing of the variables
+#' and environment available for checking. This can be helpful when you need
+#' to debug an exercise and a submission.
+#'
+#' @section Debugging exercises:
+#' 
+#' ```{r child = "man/fragments/debug_this-usage-setup.Rmd"}
+#' ```
+#' 
+#' \if{html}{
+#' The debug output will look like the following when used as described
+#' below.
+#' 
+#' \Sexpr[echo=FALSE,results=rd,stage=build]{
+#' submission <- gradethis::mock_this_exercise("# user submits\nx + 2", "x + 3", setup_exercise = "x <- 1", .label = "example")
+#' paste("\\\out{\n<blockquote>", gradethis::debug_this(submission)$message, "</blockquote>}", sep = "\n")
+#' }
+#' }
+#' 
+#' ```{r child = "man/fragments/debug_this-usage.Rmd"}
+#' ```
+#' 
+#' @examples
+#' # Suppose we have an exercise (guess the number 42). Mock a submission:
+#' submission <- mock_this_exercise(.user_code = 40, .solution_code = 11 + 31)
+#' 
+#' # Call `debug_this()` inside your *-check chunk, is equivalent to
+#' debug_this()(submission)$message
+#' 
+#' # Or you can call `debug_this()` inside a `grade_this()` call
+#' # at the point where you want to get debug feedback.
+#' grade_this({
+#'   pass_if_equal(42, "Good stuff!")
+#' 
+#'   # Find out why this is failing??
+#'   debug_this()
+#' })(submission)
+#' 
+#' # Set default `fail()` message to show debug information
+#' # (for tutorial development only!)
+#' old_opts <- options(gradethis.fail = "{debug_this()}")
+#' 
+#' grade_this({
+#'   pass_if_equal(42, "Good stuff!")
+#' 
+#'   fail()
+#' })(submission)
+#' 
+#' # default fail() will show debug until you reset gradethis.fail option
+#' options(old_opts)
+#' 
+#' @param check_env A grade checking environment. You can use
+#'   [mock_this_exercise()] to prepare a mocked exercise submission
+#'   environment. Otherwise, you don't need to use or set this argument.
+#'   
+#' @return Returns a neutral grade containing a message that includes any
+#'   and all information available about the exercise and the current 
+#'   submission. This information is all available for use within
+#'   [grade_this()].
+#' 
+#' @export
+debug_this <- function(check_env = parent.frame()) {
```

Comment:
Do you imagine `debug_this` being used in other functions? (Such as `grade_result()` or a future `grade_this_other_function()`)

Reply:
Possibly? I briefly attempted to get it to work in `grade_result()` but that's a bit trickier and given that we're deprecating it I decided not to go far down that path.

In future functions, I would expect we could adjust this function with more guards around detecting or finding the expected variables in the `check_env`.

Reply:
👍